stumpy/config.py

```diff
@@ -2,15 +2,64 @@
 # Copyright 2019 TD Ameritrade. Released under the terms of the 3-Clause BSD license.
 # STUMPY is a trademark of TD Ameritrade IP Company, Inc. All rights reserved.
 
+import warnings
+
 import numpy as np
 
-STUMPY_THREADS_PER_BLOCK = 512
-STUMPY_MEAN_STD_NUM_CHUNKS = 1
-STUMPY_MEAN_STD_MAX_ITER = 10
-STUMPY_DENOM_THRESHOLD = 1e-14
-STUMPY_STDDEV_THRESHOLD = 1e-7
-STUMPY_P_NORM_THRESHOLD = 1e-14
-STUMPY_TEST_PRECISION = 5
-STUMPY_MAX_P_NORM_DISTANCE = np.finfo(np.float64).max
-STUMPY_MAX_DISTANCE = np.sqrt(STUMPY_MAX_P_NORM_DISTANCE)
-STUMPY_EXCL_ZONE_DENOM = 4
+_STUMPY_DEFAULTS = {
+    "STUMPY_THREADS_PER_BLOCK": 512,
+    "STUMPY_MEAN_STD_NUM_CHUNKS": 1,
+    "STUMPY_MEAN_STD_MAX_ITER": 10,
+    "STUMPY_DENOM_THRESHOLD": 1e-14,
+    "STUMPY_STDDEV_THRESHOLD": 1e-7,
+    "STUMPY_P_NORM_THRESHOLD": 1e-14,
+    "STUMPY_TEST_PRECISION": 5,
+    "STUMPY_MAX_P_NORM_DISTANCE": np.finfo(np.float64).max,
+    "STUMPY_MAX_DISTANCE": np.sqrt(np.finfo(np.float64).max),
+    "STUMPY_EXCL_ZONE_DENOM": 4,
+    "STUMPY_FASTMATH": True,
+    "STUMPY_FASTMATH_FLAGS": {"nsz", "arcp", "contract", "afn", "reassoc"},
+}
+
+STUMPY_THREADS_PER_BLOCK = _STUMPY_DEFAULTS["STUMPY_THREADS_PER_BLOCK"]
+STUMPY_MEAN_STD_NUM_CHUNKS = _STUMPY_DEFAULTS["STUMPY_MEAN_STD_NUM_CHUNKS"]
+STUMPY_MEAN_STD_MAX_ITER = _STUMPY_DEFAULTS["STUMPY_MEAN_STD_MAX_ITER"]
+STUMPY_DENOM_THRESHOLD = _STUMPY_DEFAULTS["STUMPY_DENOM_THRESHOLD"]
+STUMPY_STDDEV_THRESHOLD = _STUMPY_DEFAULTS["STUMPY_STDDEV_THRESHOLD"]
+STUMPY_P_NORM_THRESHOLD = _STUMPY_DEFAULTS["STUMPY_P_NORM_THRESHOLD"]
+STUMPY_TEST_PRECISION = _STUMPY_DEFAULTS["STUMPY_TEST_PRECISION"]
+STUMPY_MAX_P_NORM_DISTANCE = _STUMPY_DEFAULTS["STUMPY_MAX_P_NORM_DISTANCE"]
+STUMPY_MAX_DISTANCE = _STUMPY_DEFAULTS["STUMPY_MAX_DISTANCE"]
+STUMPY_EXCL_ZONE_DENOM = _STUMPY_DEFAULTS["STUMPY_EXCL_ZONE_DENOM"]
+STUMPY_FASTMATH = _STUMPY_DEFAULTS["STUMPY_FASTMATH"]
+STUMPY_FASTMATH_FLAGS = _STUMPY_DEFAULTS["STUMPY_FASTMATH_FLAGS"]
+
+
+def _reset(var=None):
+    """
+    Reset the value of a configuration variable(s) to their default value(s)
+
+    Parameters
+    ----------
+    var : str, default None
+        The name of the configuration variable. If None, then all
+        configuration variables are reset to their default values.
+
+    Returns
+    -------
+    None
```

Comment:
Maybe we should add a note here to just let user know that resetting value does not necessarily mean that it becomes effective. We may also add something like this: "if it is part of njit function, the function needs to be recompiled"

Reply:
You might as well add an example section in the docstring too that shows exactly what can/should be done. 

There is one issue which is the user may not know what the default `fastmath` flags for their specific function is when calling `cache._recompile()`. Maybe it would be helpful to add a function that allows the user to query the default value? Is this ia problem? I'm just trying to think through the logic from the user's standpoint

Reply:
> You might as well add an example section in the docstring too that shows exactly what can/should be done.

Will do!!

> There is one issue which is the user may not know what the default fastmath flags for their specific function is when calling `cache._recompile()` Maybe it would be helpful to add a function that allows the user to query the default value? Is this ia problem?

I like the idea! I personally prefer to avoid exploring a codebase as much as possible. So, even if a user is advanced, they may want to easily check the default fastmath of a STUMPY's function. I think I need to use `ast`. See if I can reuse some parts of `cache.get_njit_funcs`.

Reply:
To be clear, I'm not trying to dictate anything here so please try to step through the user experience in your mind. I think it's pretty natural to call `cache._recompile(some_njit_func)` but it's rather unclear what the user should do with the `fastmath` parameter. And, in reality, there's really only three options:

1. Recompile with a truly custom set of flags
2. `fastmath=config.STUMPY_FASTMATH`
3. `fastmath=config.STUMPY_FASTMATH_FLAGS`

Maybe it's a bit weird (fishy) but you could possibly attach an attribute to each njit function:

```
njit_funcs = get_njit_funcs()
for module_name, func_name in njit_funcs:
        module = importlib.import_module(f".{module_name}", package="stumpy")
        func = getattr(module, func_name)
        func.STUMPY_FASTMATH_DEFAULTS = func.targetoptions['fastmath']
```

Okay, this looks and feels very hackish :( It would probably be something that would live in `__init__.py`. I'm starting to regret this idea...

Reply:
> To be clear, I'm not trying to dictate anything here so please try to step through the user experience in your mind

👍 

----

> I think it's pretty natural to call cache._recompile(some_njit_func) but it's rather unclear what the user should do with the fastmath parameter

This is the spectrum: 

`fastmath=False --> a set of fastmath flags --> fastmath=True`

So, if a user reaches a point where they think they should play with fastmath flags, then they can either go all they way to the right and just do `cache._recompile(..., fastmath=True)`, or they can move towards to the left of spectrum. In that case, they may want to start from the default fastmath flags of that function, and remove one or more flags from it. So, allowing users to get that default flags should help them.


@seanlaw 
I am going to take one step back here and ask: How do you feel about removing the param `fastmath` from `cache._recompile`?  I am not trying to escape from it 😄 but I am genuinely wondering whether adding this amount of complexity is worth it. From your perspective, how much will developers & users encounter a situation in which they need to play with fastmath flags?

Reply:
> From your perspective, how much will developers & users encounter a situation in which they need to play with fastmath flags?

Probably never. So, what happens to the `cache._recompile()` API? How should it be implemented if we remove the `fastmath` parameter? I think my brain immediately added the `fastmath` parameter to make it easier/more explicit but maybe there's a simpler option?

Reply:
> what happens to the cache._recompile() API? How should it be implemented if we remove the fastmath parameter?

From my perspective, it can become a `_recompile_all()` function. For a single function, one can do `func.recompile()`, and if the goal is to recompile all, then one can call `cache._recompile_all()`. 

The obvious downside is that any change in fastmath flags needs to happen outside of this function. Although one needs to do `func.targetoptions['fastmath'] = ...` every time they need to change the flags, that should not be a big problem if the chance of getting into such situation is very rare.

Reply:
@seanlaw 

> Maybe it would be helpful to add a function that allows the user to query the default value? Is this a problem?

Okay... I now think that this becomes important if the function `_recompile` is allowed to manipulate `fastmath flags` for ALL njit functions.

Suppose that a user decides to do `_recompile(func=None, fastmath=True)` and set the fastmath flags of all njit functions to `True`. How can they revert it? The function `config._reset()` just resets the value of config variables. To revert changes we need to do something like this:

```
for module_name, func_name in get_njit_funcs():
        module = importlib.import_module(f".{module_name}", package="stumpy")
        func = getattr(module, func_name)
        func.targetoptions['fastmath'] = <default value of fastmath in func>
        func.recompile()
```

Reply:
>From my perspective, it can become a _recompile_all() function. For a single function, one can do func.recompile(), and if the goal is to recompile all, then one can call cache._recompile_all().

So, what bothers me right now is that there would be an inconsistency in the `cache.py` functions. Currently, `cache._enable()` and `cache._clear()` are applied to all STUMPY njitted functions. Thus, it would be reasonable to expect that `cache._recompile()` would also be applied to all functions as well. I actually think this is fine even for our current use case (to blindly recompile all of the functions without setting any `fastmath` flags while INSIDE of the function and forcing users to set it outside of the function). This isn't a bad thing. The thing to check is whether `cache._recompile()` actually allows the `cache` files to exist. Currently, when we do:

1. `cache._enable()`
2. `stumpy.stump(T, m)`
3. `cache._get_cache()`  # This prints cached files
4. `cache._clear()`
5. `stumpy.stump(T, m)`
6. `cache._get_cache()`  # This does NOT print anything!

> Suppose that a user decides to do _recompile(func=None, fastmath=True) and set the fastmath flags of all njit functions to True. How can they revert it? The function config._reset() just resets the value of config variables. To revert changes we need to do something like this:

Exactly! That's what I identified in my head as well. The `fastmath` parameter is all-or-nothing. For users, the `fastmath` parameter is probably unnecessary. However, for you, you'll need it in order to specify removing `reassoc` for a single function. Of course, as mentioned above, you can always set the `targetoptions` for the one function yourself and then call `cache._recompile()` to recompile ALL njitted functions (even ones where you didn't change their `fastmath`) and it should be fine. It's inefficient but fine.

Reply:
>  Currently, when we do:
>
> > 1. cache._enable()
>> 2. stumpy.stump(T, m)
>> 3. cache._get_cache() # This prints cached files
>> 4. cache._clear()
>> 5. stumpy.stump(T, m)
>> 6. cache._get_cache() # This does NOT print anything!

I thought step 5 would result in writing files into cache directory! Does step 4 do something like clear caches AND disable caching? Is that an expected behaviour?


> The thing to check is whether cache._recompile() actually allows the cache files to exist

I do not know what should be the expected behaviour here. If we enable cache and clear it later, and then do `cache._recompile()`, it will write files into cache directory. From my perspective, this can be an expected behaviour. But, if the expectation is that it should behave like the provided example, then we should allow user to clear cache as you suggested before. Is that the reason you mentioned the param `clear` for `cache._recompile`?

```
import warnings
import numpy as np

from stumpy import core, cache


warnings.filterwarnings("ignore")

func = core._calculate_squared_distance

s = 3
QT = 1453258.774722079
μ_Q = -55.35781464461123
σ_Q = 650.912209452633
μ_T = -264.9540642227162
σ_T = 722.0717285148526
isconstant_Q = False
isconstant_T = False
args = (s, QT, μ_Q, σ_Q, μ_T, σ_T, isconstant_Q, isconstant_T)

cache._enable()
func(*args)
print(cache._get_cache())

cache._clear()
print(cache._get_cache())

cache._recompile(func)
print(cache._get_cache())
```

---

> and then call cache._recompile() to recompile ALL njitted functions (even ones where you didn't change their fastmath) and it should be fine. It's inefficient but fine.

We need to recompile all the njit-callers of that njit function, and the callers of those callers, and so on. So,
the efficient approach can be adding a param like `recompile_all_callers = False` and then when it is set to True, the `cache._recompile` applies the function `.recompile()` to ALL callers of the recently-changed function, and the callers of those callers, and so on. So, recompiling all BLINDLY may not be that bad given the fact that it takes a negligible time.

Reply:
> I thought step 5 would result in writing files into cache directory! Does step 4 do something like clear caches AND disable caching? Is that an expected behaviour?

I'm not sure what is happening. All I know is that the cache isn't written anymore after it has been deleted. Previously, I had only spent time enabling the cache but I don't recall what was happening with the deletion (i.e., `cache._clear()`) since it was experimental and seemed to "work" for the user. In other words, I had only stopped at step 4 previously. I only discovered this behavior the other day so it would be great if you could take a look to see if you observe the same behavior.

```
cache._enable()
func(*args)
print(cache._get_cache())
```

This should've printed something, right??

>  If we enable cache and clear it later, and then do cache._recompile(), it will write files into cache directory. From my perspective, this can be an expected behaviour.

I think I agree. Though, my brain is a little bit stuck on the idea that if you call a STUMPY njitted function then it should compile the first time it's called and, if the cache has been cleared, it should also be ready to compiled automatically the next time it is called (without needing to explicitly call `cache._recompile()`. 

Probably, the better question to answer is what does/should "enabling cache" mean? Originally, it was so that people can save a bit of time not needing to recompile some STUMPY functions each time they open a new terminal/notebook session. So, all that the user did to avoid the recompilation was:

1. open a new session/notbook
2. `cache._enable()`
3. Call some STUMPY functions, which compiled them and stored them in site-packages

The main reason why we attempted to have a `cache._clear()` was to help ensure that they didn't have some strange local compilations and the clearing would simply delete the cached files. However, I never assessed what was happening behind the `numba` scenes. For some reason, calling the same STUMPY functions after deleting the cache files didn't seem to do much (i.e., compilation time was still "fast" and the cache files didn't exist in the usual locations). So this part feels a bit unpredictable/unknown to me at this point.

If `cache._clear()` somehow also disables caching then we probably need to keep track of the cache state via `config.STUMPY_CACHE_ENABLED` (or something like that). So, by default `config.STUMPY_CACHE_ENABLED = False` and setting `cache._enable()` should also flip `config.STUMPY_CACHE_ENABLED = True`. However, when `cache._clear()` is called, we should check the state of `config.STUMPY_CACHE_ENABLED`. If `config.STUMPY_CACHE_ENABLED` is `True` then we should figure out how to allow files/functions to cached automatically on their next call. 

This begins to suggest that there must also be a `cache._disable()` function too? Is this too complicated? 

> So, recompiling all BLINDLY may not be that bad given the fact that (i) it takes a negligible time, and (ii) we may need to recompile several njit functions because of a change in one function.

Instead of having a `recompile_all_callers = False` option, I'm proposing to ALWAYS recompile everything (i.e., all functions returned by `get_njit_funcs()`). So we won't need any extra params for `cache._recompile()`. Because, at the moment, `cache._enable()` enables **all** functions to be cached and `cache._clear()` deletes **all** files. To be consistent, I feel like `cache._recompile()` should also affect **all** functions. Given that it's negligible in time, I think we should approach it this way. This then implies that the user must set all of the `fastmath` flags manually if they need to be reset. Nobody would/should ever need to do this IMHO (i.e., "this" = change/mess with STUMPY `fastmath` flags).

Reply:
Apologies if it feels like I'm being indecisive! I am thinking through this in real-time and not having all of the necessary information yet 😢

Reply:
I have [posted a question in the numba channel](https://numba.discourse.group/t/understanding-functiong-caching/2852). Hopefully, we'll get some answers

Reply:
> I only discovered this behavior the other day so it would be great if you could take a look to see if you observe the same behavior.

It is the same for me. I did the same 6 steps you provided and got not new files at the end.

> This should've printed something, right??

Right! After `cache._clear()`, neither `func(*args); print(cache._get_cache())` nor `cache._enable(); func(*args); print(cache._get_cache())` does not print anything. They should have though.

> If the cache has been cleared, it should also be ready to compiled automatically the next time it is called (without needing to explicitly call cache._recompile().

I am a newbie in the realm of cache. So, please forgive me if you notice I am just saying some meaningless stuff 😅 
IIUC, when cache is cleared, it means there is no on-disk cache... but there should still be in-memory cache. So, when we call the function again, it may re-use that in-memory cache and maybe it does not check to see if the cache directory is empty and needs to be re-filled. If they add such check, it can add some overhead. But, when we do `.recompile`, writing-cache-to-directory process is triggered again.

> I'm proposing to ALWAYS recompile everything
>  Nobody would/should ever need to do this IMHO (i.e., "this" = change/mess with STUMPY fastmath flags).

I totally agree!

> Apologies if it feels like I'm being indecisive!

We are on the same boat! 👍 This is a very nice activity for me!